### ISSUE TITLE
Clarification: NorMuon global scaling in code vs constant-RMS scaling in paper

### ISSUE THREAD

**Issue by ethaniunu** [zichongli5/NorMuon#6]
Hi, thanks for open-sourcing the NorMuon implementation and paper. I’ve been trying to line up the algorithm in the paper with the reference code, and I think there is a second discrepancy (in addition to the Nesterov vs plain momentum question).

This one is about the **global scaling of the NorMuon update** after the neuron-wise normalization.

---

### What the paper describes

In the NorMuon section, after orthogonalization and neuron-wise normalization, the paper describes a global scaling that enforces a *constant RMS* for the update.

Conceptually (in text / pseudo-code):

```text
1. Compute momentum:
   G_t  = ∇_W L(W_t)
   M_t  = β1 * M_{t-1} + (1 - β1) * G_t

2. Orthogonalize:
   O_t  = NS5(M_t)

3. NorMuon row-wise scaling:
   v_t  = β2 * v_{t-1} + (1 - β2) * meancols(O_t ⊙ O_t)   # one scalar per row
   Ō_t  = O_t / sqrt(v_t + ε)                             # row-wise division

4. Global RMS scaling:
   η̂_t = 0.2 * η * sqrt(m * n) / ||Ō_t||_F
   W_{t+1} = W_t - η * λ * W_t - η̂_t * Ō_t
```

So, after NorMuon normalization, the Frobenius norm of the update is essentially controlled by:

```text
||η̂_t * Ō_t||_F ≈ 0.2 * η * sqrt(m * n)
```

i.e. the RMS of the update is fixed as a function of LR and layer size.

---

### What the reference code does

In the reference implementation, the `normuon_update` function looks like:

```python
def normuon_update(grad, momentum, second_momentum,
                   beta=0.95, beta2=0.95, ns_steps=5, nesterov=True):

    momentum.lerp_(grad, 1 - beta)
    update = grad.lerp_(momentum, beta) if nesterov else momentum

    # Orthogonalization
    update = zeropower_via_newtonschulz5(update, steps=ns_steps).float()

    # --- NorMuon part ---
    vnorm = update.norm(dim=(-2,-1), keepdim=True)          # pre-NorMuon Frobenius norm

    v_mean = torch.mean(update * update, dim=-1, keepdim=True)
    second_momentum.lerp_(v_mean, 1 - beta2)
    step_size = 1 / second_momentum.sqrt().add_(1e-10)
    update.mul_(step_size)                                  # row-wise NorMuon scaling

    vnorm_new = update.norm(dim=(-2,-1), keepdim=True)      # post-NorMuon norm
    update.mul_(vnorm / (vnorm_new.add_(1e-10)))            # restore original vnorm
    # ---------------------

    update *= max(1, grad.size(-2) / grad.size(-1))**0.5    # aspect-ratio factor
    return update
...
p.add_(update.reshape(p.shape), alpha=-group["lr"])
```

So the actual behavior is:

1. Orthogonalize → `update`.
2. Record `vnorm = ||update||` before NorMuon normalization.
3. Apply NorMuon row-wise normalization (via `second_momentum`).
4. Compute `vnorm_new` after NorMuon normalization.
5. Rescale to *restore the original norm* (`vnorm / vnorm_new`), rather than enforcing a fixed RMS.
6. Apply an additional `sqrt(max(1, m/n))` aspect-ratio factor.
7. Multiply everything by `lr` at apply time.

This means the final Frobenius norm of the update is approximately:

```text
||ΔW||_F ≈ lr * sqrt(max(1, m/n)) * vnorm_pre
```

where `vnorm_pre` is the norm of the orthogonalized update before NorMuon scaling, and is **not** explicitly tied to `sqrt(m * n)` and the `0.2` factor as in the algorithm description.

---

### Questions

Could you clarify:

1. Is this **norm-preserving + aspect-ratio scaling** behavior (restore `vnorm` after NorMuon + `sqrt(max(1, m/n))`) the intended, “official” NorMuon implementation?
2. Were the main NorMuon results in the paper (e.g., vs Muon and AdamW on 1.1B / 5.4B) run with:

   * the constant-RMS scheme described in the text (using something like `0.2 * η * sqrt(mn) / ||Ō_t||`), or
   * the norm-preserving scheme currently in the code (restore `vnorm`, then aspect factor)?
3. If someone wants to **exactly reproduce the algorithm as written in the paper**, would you recommend:

   * modifying the implementation to use the constant-RMS scaling, or
   * sticking with the current norm-preserving behavior in the repo?

Thanks again for the great work and for sharing the code!
